### PR TITLE
[newrelic-logging] allow initcontainer configuration

### PR DESCRIPTION
**Pull request by woehrl01**
#### Is this a new chart

no

#### What this PR does / why we need it:

Adds initContainer to the nri-logging chart. This is needed to provide additional files, like external plugins to fluentbit.

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[mychartname]`)

### CODE REVIEW

File: charts/newrelic-logging/templates/daemonset.yaml [hunk 1]
```diff
@@ -44,6 +44,10 @@ spec:
       securityContext:
         {{- . | nindent 8 }}
       {{- end }}
+      {{- if .Values.initContainers }}
+      initContainers:
+{{ toYaml .Values.initContainers | indent 8 }}
+      {{- end }}
```
Comment (jsubirat) [inline charts/newrelic-logging/templates/daemonset.yaml:50]: Please also add this to the Windows daemonset (`daemonset-windows.yml`) as well. Please bear in mind that that file initializes an _array_ of Daemonsets, so you'll need to use `$.Values.initContainers` instead.

File: charts/newrelic-logging/README.md [hunk 2]
```diff
@@ -135,6 +135,7 @@ See [values.yaml](values.yaml) for the default values
...
 | `extraVolumeMounts`                                        | Additional DaemonSet volume mounts	                                                                                                                                                                                                                     | `[]`                                                                 |
 | `extraVolumes`                                             | Additional DaemonSet volumes 	                                                                                                                                                   | `[]`                                                                 |
+| `initContainers`                                             | Additional init containers 	                                                                                                                                                   | `[]`                                                                 |
```
Comment (jsubirat) [inline charts/newrelic-logging/README.md:138]:
```suggestion
| `initContainers`                                             | [Init containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/) that will be executed before the actual container in charge of shipping logs to New Relic is initialized. Use this if you are using a custom Fluent Bit configuration that requires downloading certain files inside the volumes being accessed by the log-shipping pod.	                                                                                                                                                   | `[]`                                                                 |
```

**Review by jsubirat — COMMENTED** [review]
Thank you for your contribution, added a couple of comments for you to consider.

**Comment by woehrl01** [comment]
@jsubirat Thanks for your valuable feedback! I just pushed the changes.

**Comment by woehrl01** [comment]
@jsubirat do you have any additional objections to have this merged?

**Review by jsubirat — COMMENTED** [review]
Hi @woehrl01 , sorry for this delayed review. I have one extra comment that would affect the Helm chart when being deployed on heterogeneous clusters. Could you please give it a look? Thanks.

File: charts/newrelic-logging/README.md [hunk 3]
```diff
@@ -135,6 +135,7 @@ See [values.yaml](values.yaml) for the default values
...
 | `extraVolumes`                                             | Additional DaemonSet volumes 	                                                                                                                                                   | `[]`                                                                 |
+| `initContainers`                                             | [Init containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/) that will be executed before the actual container in charge of shipping logs to New Relic is initialized. Use this if you are using a custom Fluent Bit configuration that requires downloading certain files inside the volumes being accessed by the log-shipping pod.	                                                                                                                                                   | `[]`                                                                 |
```
[the remark this answers is not in the document]
Comment (woehrl01) [reply charts/newrelic-logging/README.md:138]: Thank you @jsubirat I just pushed the changes.

**Review by bmcfeely — APPROVED** [review]
Verified that the Helm chart still works when I uncomment the `initContainers` value in `daemonset.yml`:

```
  Init Containers:
   init:
    Image:      busybox
    Port:       <none>
    Host Port:  <none>
    Command:
      sh
      -c
      echo "hello world"
    Environment:  <none>
    Mounts:       <none>
```

File: charts/newrelic-logging/Chart.yaml [hunk 4]
```diff
@@ -1,7 +1,7 @@
 apiVersion: v2
 description: A Helm chart to deploy New Relic Kubernetes Logging as a DaemonSet, supporting both Linux and Windows nodes and containers
 name: newrelic-logging
-version: 1.13.1
+version: 1.13.2
```
Comment (bmcfeely) [inline charts/newrelic-logging/Chart.yaml:4]: @woehrl01 it looks like `nri-bundle/Charts.lock` line 25 will need to be updated with this new value. 

I'm seeing a build error:
```

1 chart(s) linted, 0 chart(s) failed
Error: Chart.lock is out of sync with Chart.yaml
Error: Error linting charts: Error building dependencies for chart 'nri-bundle => (version: "5.0.11", path: "charts/nri-bundle")': Error waiting for process: exit status 1
```

Thank you!

Comment (woehrl01) [reply charts/newrelic-logging/Chart.yaml:4]: @bmcfeely  thanks for pointing that out, I just pushed the fix

**Comment by bmcfeely** [comment]
@woehrl01 the build is still complaining about the lock file. I'll need to ask around to get help, I'm not familiar with this part of the build system. My question may take about a day to get answered, I'll be back tomorrow with an answer!

**Comment by luckslovez** [comment]
> @woehrl01 the build is still complaining about the lock file. I'll need to ask around to get help, I'm not familiar with this part of the build system. My question may take about a day to get answered, I'll be back tomorrow with an answer!

I believe this needs a separate PR. When the new version of the logging chart is made available in the helm repo, the `nri-bundle` chart can be updated. Also, note that the formal way of updating a dependency in a chart is `helm dependency update` (see https://helm.sh/docs/helm/helm_dependency_update/), this way the digest and the date are updated programmatically (see [this example](https://github.com/newrelic/helm-charts/pull/937/files)).

Nevertheless, you could just let the [Renovate](https://github.com/newrelic/helm-charts#automated-version-bumps) bot do the follow up PR for you. You can always open the pr yourself though if you wish to speed up the process.

**Comment by bmcfeely** [comment]
@woehrl01 OK, I *think* I know what I'm doing now 😛 
* Please roll back the changes to `nri-bundle` (so revert changes to both`charts/nri-bundle/Chart.yaml` and `charts/nri-bundle/Chart.lock`)
* Then that should cause the linter to be happy (I think there's no way to update the nri-bundle lock file of to get the new version of newrelic-logging, since newrelic-logging hasn't yet been "published") 
* Then I'll merge this PR
* Then I'll create a new PR to update `nri-bundle` (from what I understand, `newrelic-logging` needs to be updated first, since it's in the same repository as `nri-bundle`)

If you are feeling "why are you making me do so many changes?" (a) I sympathize, thank you for your patience 😄 , and (b) just let me know if you'd prefer for me to do that rollback -- I can make a new PR from your changes and roll back the nri-bundle changes myself (but I'm not sure don't know if I know how to get the commits from your fork, so it might look like I made this good change 😄). lmk!

**Comment by woehrl01** [comment]
@bmcfeely no worries at all! I'll make the changes as soon as I'm at my desk

**Comment by woehrl01** [comment]
@bmcfeely Reverted the changes

**Comment by bmcfeely** [comment]
Creating a new PR at #1085 so that @woehrl01 doesn't have to keep making changes (conflict resolution, merges to get build changes)

**Comment by bmcfeely** [comment]
@woehrl01 these changes are now in the newrelic-logging chart (see #1085) -- when #1082 is merged (hopefully in the next few day), the nri-bundle chart will get updated to pick up that change as well. Thank you for your patience!

**Pull request closed without merging**